keanu-project/src/main/java/io/improbable/keanu/templating/SequenceItem.java

```diff
@@ -121,4 +124,85 @@ public SequenceItem withExtraEntries(Map<VertexLabel, Vertex<?>> extraEntries) {
+        DoubleProxyVertex newVertex = new DoubleProxyVertex(shape, vertexLabel);
+        this.add(newVertex);
+        return newVertex;
+    }
+
+    /**
+     * This method creates a {@link ProxyVertex} and adds it to the sequence item
+     * It also adds this vertex to the sequence item to represent the vertex with this label in the previous sequence item.
+     * @param label the label of the corresponding vertex from the previous item.
+     * @return a newly created {@link ProxyVertex}
+     */
+    public IntegerProxyVertex addIntegerProxyFor(VertexLabel label) {
+        VertexLabel vertexLabel = SequenceBuilder.proxyLabelFor(label);
+        IntegerProxyVertex newVertex = new IntegerProxyVertex(vertexLabel);
+        this.add(newVertex);
+        return newVertex;
+    }
+
+    /**
+     * This method creates a {@link ProxyVertex} and adds it to the sequence item
+     * It also adds this vertex to the sequence item to represent the vertex with this label in the previous sequence item.
+     * @param label the label of the corresponding vertex from the previous item.
+     * @param shape the shape of the corresponding vertex from the previous item.
+     * @return a newly created {@link ProxyVertex}
+     */
+    public IntegerProxyVertex addIntegerProxyFor(VertexLabel label, long[] shape) {
+        VertexLabel vertexLabel = SequenceBuilder.proxyLabelFor(label);
+        IntegerProxyVertex newVertex = new IntegerProxyVertex(shape, vertexLabel);
+        this.add(newVertex);
+        return newVertex;
+    }
+
+    /**
+     * This method creates a {@link ProxyVertex} and adds it to the sequence item
+     * It also adds this vertex to the sequence item to represent the vertex with this label in the previous sequence item.
+     * @param label the label of the corresponding vertex from the previous item.
+     * @return a newly created {@link ProxyVertex}
+     */
+    public BooleanProxyVertex addBooleanProxyFor(VertexLabel label) {
+        VertexLabel vertexLabel = SequenceBuilder.proxyLabelFor(label);
+        BooleanProxyVertex newVertex = new BooleanProxyVertex(vertexLabel);
+        this.add(newVertex);
+        return newVertex;
+    }
+
+    /**
+     * This method creates a {@link ProxyVertex} and adds it to the sequence item
+     * It also adds this vertex to the sequence item to represent the vertex with this label in the previous sequence item.
+     * @param label the label of the corresponding vertex from the previous item.
+     * @param shape the shape of the corresponding vertex from the previous item.
+     * @return a newly created {@link ProxyVertex}
+     */
+    public BooleanProxyVertex addBooleanProxyFor(VertexLabel label, long[] shape) {
+        VertexLabel vertexLabel = SequenceBuilder.proxyLabelFor(label);
+        BooleanProxyVertex newVertex = new BooleanProxyVertex(shape, vertexLabel);
+        this.add(newVertex);
+        return newVertex;
+    }
```

Comment:
you can collapse these into a single generic method:

    public <T extends Vertex<?>> T addProxyFor(VertexLabel label, Function<VertexLabel, T> factoryMethod) {
        VertexLabel proxyLabel = SequenceBuilder.proxyLabelFor(label);
        T newVertex = factoryMethod.apply(proxyLabel);
        this.add(newVertex);
        return newVertex;
    }

Then you can call it like:

    addProxyFor(label, DoubleProxyVertex::new);

Or: 

    addProxyFor(label, (proxyLabel) -> new IntegerProxyVertex(shape, proxyLabel));

etc.

Reply:
@migwellian I like this - very elegant. In python, do you think it would be better to remain having three different methods for each type or is there a similar way of wrapping them into one function?

Reply:
Actually to be honest I'm don't really know how to pass references to methods via Py4J at all